shared/middleware/apps_api.rb

```diff
@@ -249,24 +249,24 @@ class AppsApi < Sinatra::Base
   end
 
   #
-  # PATCH (PUT, POST) /v3/apps/<app-id>/[user-]tables/<table-name>/<row-id>
+  # PATCH (PUT, POST) /v3/apps/<app-id>/<endpoint>-tables/<table-name>/<row-id>
   #
   # Update an existing row.
   #
-  post %r{/v3/apps/([^/]+)/(shared|user)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
+  post %r{/v3/apps/([^/]+)/(shared|user|readonly)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
     unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
     unsupported_media_type unless request.content_charset.to_s.downcase == 'utf-8'
 
-    value = Table.new(app_id, storage_id(endpoint), table_name).update(id, JSON.parse(request.body.read), request.ip)
+    value = Table.new(app_id, storage_id(endpoint), table_name, my_storage_id).update(id, JSON.parse(request.body.read), request.ip)
 
     dont_cache
     content_type :json
     value.to_json
   end
-  patch %r{/v3/apps/([^/]+)/(shared|user)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
+  patch %r{/v3/apps/([^/]+)/(shared|user|readonly)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
     call(env.merge('REQUEST_METHOD'=>'POST'))
   end
-  put %r{/v3/apps/([^/]+)/(shared|user)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
+  put %r{/v3/apps/([^/]+)/(shared|user|readonly)-tables/([^/]+)/(\d+)$} do |app_id, endpoint, table_name, id|
     call(env.merge('REQUEST_METHOD'=>'POST'))
   end
 
```

Comment:
This file is good, as-is, let's get it checked in.

But next time, since it will be the fourth kind of table, if the only substantial change is adding to `(shared|user|readonly)` then it may we worth finding a way to use a list to generate the variations, e.g:

```
[
  'shared',
  'user',
  'readonly',
].each do |endpoint|
  # INSERT ENTIRE CURRENT FILE IN HERE, modify functions, e.g. the final one becomes:

    put %r{/v3/apps/([^/]+)/#{endpoint}-tables/([^/]+)/(\d+)$} do |app_id, table_name, id|
     call(env.merge('REQUEST_METHOD'=>'POST'))
   end
end
```